salt/_modules/metalk8s_volumes.py

```diff
@@ -0,0 +1,430 @@
+    """Check if the backing storage device exists for the given volume.
+
+    Args:
+        name (str): volume name
+
+    Returns:
+        bool: True if the backing storage device exists, otherwise False
+
+    CLI Example:
+
+    .. code-block:: bash
+
+        salt '*' metalk8s_volumes.exists example-volume
+    """
+    return _get_volume(name).exists
+
+
+def create(name):
+    """Create the backing storage device for the given volume.
+
+    Args:
+        name (str): volume name
+
+    Returns:
+        None
+
+    CLI Example:
+
+    .. code-block:: bash
+
+        salt '*' metalk8s_volumes.create example-volume
+    """
+    return _get_volume(name).create()
+
+
+def is_finalized(name):
+    """Check if the backing storage device is finalized for the given volume.
+
+    Args:
+        path (str): path of the sparse file
+
+    Returns:
+        bool: True if the backing storage device is finalized, otherwise False
+
+    CLI Example:
+
+    .. code-block:: bash
+
+        salt '*' metalk8s_volumes.is_finalized example-volume
+    """
+    return _get_volume(name).is_finalized
+
+
+def finalize(name):
+    """Finalize the backing storage device setup of the given volume.
+
+    Args:
+        name (str): volume name
+
+    Returns:
+        str: path to the associated loop device
+
+    CLI Example:
+
+    .. code-block:: bash
+
+        salt '*' metalk8s_volumes.finalize example-volume
+    """
+    return _get_volume(name).finalize()
+
+
+def is_formatted(name):
+    """Check if the given volume is formatted.
+
+    Args:
+        name (str): volume name
+
+    Returns:
+        bool: True if the volume is already formatted, otherwise False
+
+    CLI Example:
+
+    .. code-block:: bash
+
+        salt '*' metalk8s_volumes.volume_is_formatted example-volume
+    """
+    return _get_volume(name).is_formatted
+
+
+def format(name):
+    """Format the given volume.
+
+    Args:
+        name (str): volume name
+
+    Returns:
+        None
+
+    CLI Example:
+
+    .. code-block:: bash
+
+        salt '*' metalk8s_volumes.format example-volume
+    """
+    _get_volume(name).format()
+
+
+# Volume {{{
+
+
+class Volume(object):
+    """Volume interface."""
+
+    __metaclass__ = abc.ABCMeta
+
+    def __init__(self, volume):
+        self._volume = volume
+
+    @abc.abstractproperty
+    def exists(self):
+        """Does the backing storage device exists?"""
+        return
+
+    @abc.abstractmethod
+    def create(self):
+        """Create the backing storage device."""
+        return
+
+    @abc.abstractproperty
+    def is_finalized(self):
+        """Check if the backing storage device setup is finalized."""
+        return
+
+    @abc.abstractmethod
+    def finalize(self):
+        """Finalize the setup of the backing storage device."""
+        return
+
+    @abc.abstractproperty
+    def path(self):
+        """Path to the backing device."""
+        return
+
+    @property
+    def is_formatted(self):
```

Comment:
A case I believe we're not handling right now, and should do (likely by erroring out): what if the volume exists (and has the expected size and such), it is formatted, but its FS UUID is *not* the UUID we expect it to have?

Reply:
Maybe `is_formatted` should only check the UUID in fact: that's the proof that we have formatted it ourself.

Checking the FS type can be dangerous no?
Imagine we have a StorageClass `Foo` with `fsType` set to `xfs`, we format the device as described.
Later someone change the `fsType` of the `StorageClass` (I think you can't edit `parameters` of a `StorageClass` but then simply delete/recreate a new with the same name…)  to `ext4`.
Now at the next call for this volume (like when you reboot) we will refresh the pillar, store the new version of the StorageClass and see `xfs != ext4` => not formatted.
Then we go to format => "ho it's already formatted so we bail out".
No data loss (thanks to the safeguard we added after the previous reviews) but an ugly failure.

Or I'm missing something and this scenario can't happen?

Reply:
So I think I'll rewrite `is_formatted` to check the UUID only:
- they match => nothing to do
- they mismatch => either we don't have one (so we should go format) or we have another one (we go format but then we detect the existing FS and bail out => we're safe).

Reply:
Done